tests/testthat/test-generate_data.R

```diff
@@ -381,3 +381,156 @@ test_that("genDummy works.", {
+  
+  
+  # if startProb defined, check it has length == number of matrix rows
+  mat6 <- t(matrix(c(0.7, 0.2, 0.1, 0.5, 0.3, 0.2, 0.0, 0.1, 0.9), nrow=3,ncol=3))
+  expect_error(genMarkov(n=10, transMat = mat6, chainLen = 5, wide = TRUE, startProb = ".7;.3"), class = "simstudy::lengthMismatch")
+  
+  set.seed(oldSeed)
+  
+})
+
+test_that("genMarkov works.", {
+  oldSeed <- .Random.seed
+  set.seed(23456)
+  
+  
+  # wide, not startProb
+  ## pk
+  matr <- t(matrix(c(0.5, 0.5, 0.0, 0.0,
+                     0.15, 0.5, 0.35, 0.0,
+                     0.0, 0.35, 0.5, 0.15,
+                     0.0, 0.0, 0.5, 0.5), nrow=4,ncol=4))
+  gm1 <- genMarkov(n=500, transMat = matr, chainLen = 250, wide = TRUE)
+  totals <- table(gm1[, "S250"])
+  ratios <- totals/500
+  
+  expected_v <- c(0.115, 0.3833333, 0.3833333, 0.115)
+  
+  expect_silent(ratios[1] >= expected_v[1] - 0.005 && ratios[1] <= expected_v[1] + 0.005)
+  expect_silent(ratios[2] >= expected_v[2] - 0.005 && ratios[2] <= expected_v[2] + 0.005)
+  expect_silent(ratios[3] >= expected_v[3] - 0.005 && ratios[3] <= expected_v[3] + 0.005)
+  expect_silent(ratios[4] >= expected_v[4] - 0.005 && ratios[4] <= expected_v[4] + 0.005)
+  
+  
+  ## correct number of events gen
+  expect_equal(ncol(gm1), 251)
+  
+  ## number of categories == dimensions of transistion matrix
+  for(i in 1:4) {
+    expect_equal(any(gm1 == i), TRUE)
+  }
+  
+  # wide, startProb
+  ## pk
+  set.seed(23456)
+  gm2 <- genMarkov(n=500, transMat = matr, chainLen = 250, wide = TRUE, startProb = "0.65;0.25;0.05;0.05")
+  totals <- table(gm2[, "S250"])
+  ratios <- totals/500
+  
+  expected_v <- c(0.115, 0.3833333, 0.3833333, 0.115)
+  
+  expect_silent(ratios[1] >= expected_v[1] - 0.005 && ratios[1] <= expected_v[1] + 0.005)
+  expect_silent(ratios[2] >= expected_v[2] - 0.005 && ratios[2] <= expected_v[2] + 0.005)
+  expect_silent(ratios[3] >= expected_v[3] - 0.005 && ratios[3] <= expected_v[3] + 0.005)
+  expect_silent(ratios[4] >= expected_v[4] - 0.005 && ratios[4] <= expected_v[4] + 0.005)
+  
+  ## correct number of events gen
+  expect_equal(ncol(gm2), 251)
+  
+  ## number of categories == dimensions of transistion matrix
+  for(i in 1:4) {
+    expect_equal(any(gm2 == i), TRUE)
+  }
+  
+  
+  # not wide == wide
+  set.seed(23456)
+  gm1_nw <- genMarkov(n=500, transMat = matr, chainLen = 250, wide = FALSE)
+  
+  set.seed(23456)
+  gm2_nw <- genMarkov(n=500, transMat = matr, chainLen = 250, wide = FALSE, startProb = "0.65;0.25;0.05;0.05")
+  
+  check_equal <- function (gm_not_wide, gm_wide) {
+    rand_id <- sample(1:500, 1)
+    rand_state <- sample(1:250, 1)
+    gmnw <- gm_not_wide[((250 * (rand_id - 1)) + rand_state), 3]
+    gmw <- gm_wide[rand_id, (rand_state + 1), with = FALSE]
+    colnames(gmnw) <- c("x")
+    colnames(gmw) <- c("x")
+    
+    expect_equal(gmnw, gmw)
+  }
+  
+  for (i in 1:5) {
+    check_equal(gm1_nw, gm1)
+    check_equal(gm2_nw, gm2)
+  }
+  
+  set.seed(oldSeed)
+  
+})
```

Comment:
I like what you did conceptually, but I changed the coding a bit so that it is again less hard coded.

```
set.seed(23456)
gm1_w <- genMarkov(n=nind, transMat = matr, chainLen = nchain, wide = TRUE)

set.seed(23456)
gm2_w <- genMarkov(n=nind, transMat = matr, chainLen = nchain, wide = TRUE, startProb = "0.65;0.25;0.05;0.05")

check_equal <- function (gm_not_wide, gm_wide) {
  rand_id <- sample(n, 1)
  rand_state <- sample(l, 1)
  gmnw <- gm_not_wide[id == rand_id & period == rand_state, state]
  gmw <- gm_wide[rand_id, rand_state + 1, with = FALSE]
  
  expect_equal(as.numeric(gmnw), as.numeric(gmw))
}

for (i in 1:5) {
  check_equal(gm1, gm1_w)
  check_equal(gm2, gm2_w)
}
```

Reply:
I see I'll put your stuff in. I want to confirm that I understand what you're doing in a few places. You call expect_silent a few times when I think you're looking to check whether an expression is True, for example expect_silent(all(prop[, dif <0.2])). But even if this is False, expect_silent will pass. Is this what you're expecting or am I misunderstanding this.

Reply:
You're right - in that case, it should be `expect_true`.